solarforecastarbiter/metrics/tests/test_preprocessing.py

```diff
@@ -0,0 +1,154 @@
+        interval_value_type='instantaneous',
+        interval_length=pd.Timedelta(fx_series.index.freq),
+        interval_label=interval_label,
+        issue_time_of_day=datetime.time(hour=5),
+        lead_time_to_start=pd.Timedelta('1h'),
+        run_length=pd.Timedelta('12h')
+    )
+    fx_obs = datamodel.ForecastObservation(forecast=forecast,
+                                           observation=observation)
+
+    # Use local tz
+    local_tz = f"Etc/GMT{int(time.timezone/3600):+d}"
+
+    data = {
+        fx_obs.forecast: fx_series,
+        fx_obs.observation: obs_series
+    }
+    result = preprocessing.resample_and_align(fx_obs, data, local_tz)
+
+    # Localize datetimeindex
+    expected_dt = expected_dt.tz_convert(local_tz)
+
+    pd.testing.assert_index_equal(result.forecast_values.index,
+                                  result.observation_values.index,
+                                  check_categorical=False)
+    pd.testing.assert_index_equal(result.observation_values.index,
+                                  expected_dt,
+                                  check_categorical=False)
+
+
+@pytest.mark.parametrize('fx0', [
+    pd.Series(index=pd.DatetimeIndex([], name='timestamp'), name='value'),
+    THREE_HOUR_SERIES
+])
+@pytest.mark.parametrize('fx1', [
+    pd.Series(index=pd.DatetimeIndex([], name='timestamp'), name='value'),
+    THREE_HOUR_SERIES
+])
+@pytest.mark.parametrize('obs', [
+    pd.DataFrame(index=pd.DatetimeIndex([], name='timestamp'),
+                 columns=['value', 'quality_flag']),
+    pd.DataFrame({'value': [1., 2., 3.],
+                  'quality_flag': [OK, OK, OK]},
+                 index=THREE_HOURS),
+    pd.DataFrame(index=pd.DatetimeIndex([], name='timestamp'),
+                 columns=['value', 'quality_flag']),
+    pd.DataFrame({'value': [1., 2., 3.],
+                  'quality_flag': [CL_UF, CL_UF, CL_UF]},
+                 index=THREE_HOURS),
+    pd.DataFrame(index=pd.DatetimeIndex([], name='timestamp'),
+                 columns=['value', 'quality_flag']),
+    pd.DataFrame({'value': [1., 2., 3.],
+                  'quality_flag': [CSE_NT, CSE, OK]},
+                 index=THREE_HOURS),
+])
+@pytest.mark.parametrize('handle_func', [preprocessing.exclude])
+def test_apply_validation(report_objects, fx0, fx1, obs, handle_func):
+    report, obs_model, fx0_model, fx1_model = report_objects
+    data = {
+        obs_model: obs,
+        fx0_model: fx0,
+        fx1_model: fx1
+    }
+    result = preprocessing.apply_validation(data,
+                                            report.filters[0],
+                                            handle_func)
+
+    # Check length and timestamps of observation
+    assert len(obs[obs.quality_flag.isin([OK, CSE])]) == \
+        len(result[obs_model])
+    if not result[obs_model].empty:
+        assert obs[obs.quality_flag.isin([OK, CSE])].index.equals(
+            result[obs_model].index)
+
```

Comment:
certainly doesn't hurt to check the forecast series were returned as well

Reply:
Added checks to make sure the forecast values are unchanged.